Description: Ignore LUBSAN sections
 The 5.15 Ubuntu kernel has strange linker behavior when built, resulting
 in a situation where multiple LUBSAN data sections are being marked
 changed, but are not selected for inclusion. It is not possible to mark
 these sections as ignored using macros, so we must ignore them in 
 kpatch-build instead.
Author: Benjamin M Romer <benjamin.romer@canonical.com>
Last-Update: 2022-05-24
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/kpatch-build/create-diff-object.c
+++ b/kpatch-build/create-diff-object.c
@@ -563,6 +563,12 @@
 {
 	struct section *sec1 = sec, *sec2 = sec->twin;
 
+        /* short-circuit Lubsan sections, since we will drop them */
+        if (strstr(sec->name,"Lubsan")) {
+                sec->status = SAME;
+                goto out;
+        }
+
 	/* Compare section headers (must match or fatal) */
 	if (sec1->sh.sh_type != sec2->sh.sh_type ||
 	    sec1->sh.sh_flags != sec2->sh.sh_flags ||
@@ -1522,6 +1528,14 @@
 				continue;
 
 			/*
+			 * Skip LUBSAN sections.
+			 */
+			if (strstr(rela->sym->name,"Lubsan")) {
+				log_debug("skipping %s\n", rela->sym->name);
+				continue;
+			}
+
+			/*
 			 * Replace references to bundled sections with their
 			 * symbols.
 			 */
@@ -1652,6 +1666,9 @@
 	int errs = 0;
 
 	list_for_each_entry(sec, &kelf->sections, list) {
+		if (sec->ignore)
+			continue;
+
 		if (sec->status == CHANGED && !sec->include) {
 			log_normal("changed section %s not selected for inclusion\n",
 				   sec->name);
@@ -2723,6 +2740,19 @@
 			sec->ignore = 1;
 	}
 
+	/* Ignore and kick out Lubsan_data sections not selected for inclusion */
+	list_for_each_entry(sec, &kelf->sections, list) {
+		if (strstr(sec->name,"Lubsan")) {
+			log_normal("ignoring Lubsan section: %s\n", sec->name);
+			ignoresec = sec;
+			if (is_rela_section(ignoresec))
+				ignoresec = ignoresec->base;
+			ignoresec->ignore = 1;
+			if (ignoresec->twin)
+				ignoresec->twin->ignore = 1;
+		}
+	}
+
 	sec = find_section_by_name(&kelf->sections, ".kpatch.ignore.sections");
 	if (!sec)
 		return;
